File: Medium/973. K Closest points to Origin/solution.java
class Solution {
    public int[][] kClosest(int[][] points, int K) {
        int[][] ans = new int[K][2];
        if (points == null || points.length == 0) {
            return ans;
        }
        Arrays.sort(points, new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                return a[0] * a[0] + a[1] * a[1] - b[0] * b[0] - b[1] * b[1];
            }
        });
        for (int i = 0; i < K; i++) {
            ans[i] = points[i];
        }
        return ans;
    }
}

File: Medium/973. K Closest points to Origin/solution.py
class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:

        self.quickSelect(points, 0, len(points) - 1, k)
        return points[:k]

    def quickSelect(self, points: List[List[int]], left: int, right: int, target: int) -> None:

        if left < right:

            pivot = self.partition(points, left, right)
            if pivot == target:
                return
            
            if pivot > target:
                self.quickSelect(points, left, pivot - 1, target)
            else:
                self.quickSelect(points, pivot + 1, right, target)
    

    def partition(self, points: List[List[int]], left: int, right: int) -> int:

        pivot = points[right]
        curr = left

        for i in range(left, right):
            point = points[i]
            if point[0] **2 + point[1] **2 < pivot[0] **2 + pivot[1] ** 2:
                points[curr], points[i] = points[i], points[curr]
                curr += 1
        
        points[curr], points[right] = points[right], points[curr]
        return curr

